datafusion/optimizer/src/decorrelate_where_in.rs

```diff
@@ -121,80 +121,95 @@ impl OptimizerRule for DecorrelateWhereIn {
     }
 }
 
+/// Optimize the where in subquery to left-anti/left-semi join.
+/// If the subquery is a correlated subquery, we need extract the join predicate from the subquery.
+///
+/// For example, given a query like:
+/// `select t1.a, t1.b from t1 where t1 in (select t2.a from t2 where t1.b = t2.b and t1.c > t2.c)`
+///
+/// The optimized plan will be:
+///
+/// Projection: t1.a, t1.b                                                                                                                                                                                                                                                                                       |
+///   LeftSemi Join:  Filter: t1.a = __correlated_sq_1.a AND t1.b = __correlated_sq_1.b AND t1.c > __correlated_sq_1.c                                                                                                                                                                                           |
+///     TableScan: t1                                                                                                                                                                                                                                                                                            |
+///     SubqueryAlias: __correlated_sq_1                                                                                                                                                                                                                                                                         |
+///       Projection: t2.a AS a, t2.b, t2.c                                                                                                                                                                                                                                                                      |
+///         TableScan: t2                                                                                                                                                                                                                                                                                        |
+///
```

Comment:
```suggestion
/// The optimized plan will be:
///
/// ```text
/// Projection: t1.a, t1.b
///   LeftSemi Join:  Filter: t1.a = __correlated_sq_1.a AND t1.b = __correlated_sq_1.b AND t1.c > __correlated_sq_1.c
///     TableScan: t1
///     SubqueryAlias: __correlated_sq_1
///       Projection: t2.a AS a, t2.b, t2.c
///         TableScan: t2
/// ```
```